src/Geometry/CubeGenerator.cpp

```diff
@@ -225,137 +237,118 @@ void seissol::geometry::CubeGenerator::cubeGenerator(unsigned int numCubes[4],
-  //      checkNcError(nc_var_par_access(ncFile, ncVarBndElemLocalIds, NC_COLLECTIVE));
+  int masterRank;
+  unsigned int groupSize = 1;
+#ifdef USE_MPI
+  groupSize = utils::Env::get<unsigned int>("SEISSOL_NETCDF_GROUP_SIZE", 1);
+  if (nProcs % groupSize != 0)
+    logError() << "#Processes must be a multiple of the group size" << groupSize;
+
+  int master = (rank / groupSize) * groupSize;
+  MPI_Comm commMaster;
+  MPI_Comm_split(
+      seissol::MPI::mpi.comm(), rank % groupSize == 0 ? 1 : MPI_UNDEFINED, rank, &commMaster);
+
+  masterRank = -1;
+
+  if (commMaster != MPI_COMM_NULL) {
+    MPI_Comm_rank(commMaster, &masterRank);
+  }
+#else  // USE_MPI
+  masterRank = rank; // = 0;
+#endif // USE_MPI
+  size_t bndSize = -1;
+  size_t bndElemSize = -1;
+
+  bool hasGroup = false;
+
+  int* sizes = 0L;
+  int maxSize = 0;
+  if (masterRank == 0) {
+    // Get important dimensions
+    size_t partitions = numPartitions[3];
+
+    if (partitions != static_cast<unsigned int>(nProcs))
+      logError() << "Number of partitions in netCDF file does not match number of MPI ranks.";
+
+    bndSize = 6;
+    bndElemSize = *std::max_element(numBndElements, numBndElements + 3);
+  }
+
+#ifdef USE_MPI
+  // Broadcast boundary sizes
+  unsigned long buf[2] = {bndSize, bndElemSize};
+  MPI_Bcast(buf, 2, MPI_UNSIGNED_LONG, 0, seissol::MPI::mpi.comm());
+  bndSize = buf[0];
+  bndElemSize = buf[1];
+#endif // USE_MPI
+
+  if (masterRank >= 0) {
+    // Elements
+    sizes = new int[groupSize];
+
+    for (int i = groupSize - 1; i >= 0; i--) {
+      int size = numElemPerPart[3];
+
+      sizes[i] = size;
+      if (i != 0) {
+#ifdef USE_MPI
+        // Forward size (except the own)
+        MPI_Send(&size, 1, MPI_INT, i + rank, 0, seissol::MPI::mpi.comm());
+#else  // USE_MPI
+        assert(false);
+#endif // USE_MPI
+      }
+
+      maxSize = std::max(maxSize, size);
+    }
+  } else {
+#ifdef USE_MPI
+    sizes = new int[1];
+    MPI_Recv(sizes, 1, MPI_INT, master, 0, seissol::MPI::mpi.comm(), MPI_STATUS_IGNORE);
+
+    maxSize = sizes[0];
+#else  // USE_MPI
+    assert(false);
+#endif // USE_MPI
+  }
+
+#ifdef USE_MPI
+  // Broadcast group information
+  int iHasGroup = hasGroup;
+  MPI_Bcast(&iHasGroup, 1, MPI_INT, 0, seissol::MPI::mpi.comm());
+  hasGroup = iHasGroup != 0;
+#endif // USE_MPI
 
   int writes_done = 0;
-  loadBar(writes_done, netcdf_writes);
+  loadBar(writes_done, netcdf_writes, rank);
+
+  m_elements.resize(sizes[0]);
 
-  // Elements
-  int* elemSize = new int[numPartitions[3]];
-  std::fill(elemSize, elemSize + numPartitions[3], numElemPerPart[3]);
-  checkNcError(nc_put_var_int(ncFile, ncVarElemSize, elemSize));
-  delete[] elemSize;
   writes_done++;
-  loadBar(writes_done, netcdf_writes);
+  loadBar(writes_done, netcdf_writes, rank);
 
   std::vector<CubeVertex> vertices;
   vertices.resize(numElemPerPart[3] * 4);
+
+#ifdef _OPENMP
 #pragma omp parallel for
```

Comment:
You could add `schedule(static)` here (same for all other OMP fors).

Secondly, you can add a `collapse(3)` here (however, that may not work everywhere; especially if there are `++`es or so between the iterations).